usersession/userd/launcher.go

```diff
@@ -170,6 +185,216 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface. The desktopFileID is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *Launcher) OpenDesktopEntryEnv(desktopFileID string, env []string, sender dbus.Sender) *dbus.Error {
+	desktopFile, err := desktopFileIDToFilename(existsOnFileSystem, desktopFileID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	err = verifyDesktopFileLocation(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := parseExecCommand(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		if !strutil.ListContains(allowedEnvVars, strings.SplitN(e, "=", 2)[0]) {
+			return dbus.MakeFailedError(fmt.Errorf("Supplied environment variable %q is not allowed", e))
+		}
+
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	// XXX: this avoids defunct processes but causes userd to persist
+	// until all children are gone (currently, this is not a problem since
+	// userd is long running once started)
+	go cmd.Wait()
+
+	if cmd.Run() != nil {
```

Comment:
[`Cmd.Run`][1] waits for the subprocess to exit.  Unless we are making use of an intermediate child we expect to exit quickly, this should be `cmd.Start()`.

[1]: https://golang.org/pkg/os/exec/#Cmd.Run

Reply:
Using `systemd-run` resolves this